penguin/tools/runtime.py

```
@@ -124,6 +208,132 @@ def hash_tool_output(output: Any) -> str:
     return hashlib.sha256(output_text.encode()).hexdigest()
 
 
+def hash_tool_arguments(arguments: Any) -> str:
+    """Return a stable hash for normalized tool arguments."""
+
+    return hashlib.sha256(
+        _stable_json(_normalize_arguments(arguments)).encode()
+    ).hexdigest()
+
+
+def _preview_text(value: Any, *, max_chars: int) -> str:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `value`

(ANN401)